Fix hdf5

jbrown725 opened the pull request:
Change Hdf5Recorder to initialize `burstinessHist_` and `spikesHistory_` as arrays large enough to hold data for multi-epoch simulations. Closes #234.

Incremental HDF5 file writing from BrainGrid was already implemented in d4dc1240a9fb1f8332466846f6012b2a93f1e887. I verified that it was writing readable files using the `h5dump --header` command. I also checked the output file of a simulation in progress and saw that the file size increases after every epoch.

jbrown725 commented:
It looks like some commits from my regression testing branch made their way into this PR. Still learning the finer points of git.

stiber commented on .gitignore:
```diff
@@ -80,3 +80,5 @@ Output/Debug/*.txt
 # Result files
 Output/Results/*.xml
 Output/Results/*.h5
+Testing/RegressionTesting/TestOutput/*.xml
+Testing/RegressionTesting/TestOutput/*.h5
```
Maybe just the whole directories (`Results` and `TestOutput`)?

stiber commented on Simulator/Recorders/Hdf5Recorder.cpp:
```diff
@@ -169,20 +171,20 @@ void Hdf5Recorder::initDataSet()
     // allocate and initialize data memories
-    burstinessHist_ = new int[static_cast<int>(Simulator::getInstance().getEpochDuration())];
-   spikesHistory_ = new int[static_cast<int>(Simulator::getInstance().getEpochDuration() * 100)];
-    memset(burstinessHist_, 0, static_cast<int>(Simulator::getInstance().getEpochDuration() * sizeof(int)));
-    memset(spikesHistory_, 0, static_cast<int>(Simulator::getInstance().getEpochDuration() * 100 * sizeof(int)));
+    burstinessHist_ = new int[static_cast<int>(simulator.getEpochDuration() * simulator.getNumEpochs())];
+    spikesHistory_ = new int[static_cast<int>(simulator.getEpochDuration() * simulator.getNumEpochs() * 100)];
+    memset(burstinessHist_, 0, static_cast<int>(simulator.getEpochDuration() * simulator.getNumEpochs() * sizeof(int)));
+    memset(spikesHistory_, 0, static_cast<int>(simulator.getEpochDuration() * simulator.getNumEpochs() * 100 * sizeof(int)));
```
I assume this is the place where the epoch data buffers were reallocated to be much larger. If that isn't the case in BrainGrid, then it must be the case that there is code in BG that avoids the problem we're seeing in Graphitti. Rather than hack this solution, we should find where the missing pieces are in Graphitti and fix them. Otherwise, not only do these become enormous for long simulations, but there will be incompatibilities with the GPU code.

stiber replied:
Given fixes in following commits, looks good.

stiber reviewed (changes requested):
Hopefully, my comments are coherent. I'm a bit tired this night. 😪

jbrown725 commented:
For Hdf5Recorder:
- Reverted `burstinessHist_` and `spikesHistory_` arrays to original lengths 
- Added back a term in the calculation of idx1 and idx2 that was removed in b82c9d33e3cc3dfba6467b9aff6e5eedce309d25. This term keeps the indices within the bounds of the arrays.

For Hdf5GrowthRecorder:
- Shortened the dimensions of `ratesHistory` and `radiiHistory` dataspaces
    - Their dimensions were made bigger in 0b9f08e1478a58a738c5e10aa625d8e53a45a27c but I'm not sure why. They had a lot of empty space in them

HDF5 files now output without hitting any assertions or seg faults. I verified the outputs by converting them to xml with the `h5dump --xml` command and diffing BrainGrid vs Graphitti outputs. They're all identical now.
